cpp/generate_year_correction_list.cc

```diff
@@ -0,0 +1,118 @@
+/** \brief Utility for displaying various bits of info about a collection of MARC records.
+ *  \author Dr. Johannes Ruscheinski (johannes.ruscheinski@uni-tuebingen.de)
+ *
+ *  \copyright 2020 Universitätsbibliothek Tübingen.  All rights reserved.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <stdexcept>
+#include <unordered_set>
+#include <vector>
+#include <cstdio>
+#include <cstdlib>
+#include "FileUtil.h"
+#include "MARC.h"
+#include "MiscUtil.h"
+#include "RegexMatcher.h"
+#include "util.h"
+
+
+namespace {
+
+
+[[noreturn]] void Usage() {
+    ::Usage("marc_data superior_ppn_list");
+}
+
+
+void ProcessRecords(MARC::Reader * const marc_reader, File * const output) {
+    unsigned record_count(0), matched_articles_count(0);
+    std::unordered_set<std::string> superior_ppns;
+    while (const MARC::Record record = marc_reader->read()) {
+        ++record_count;
+
+        if (not record.isArticle())
+            continue;
+
+        bool found_iSWA(false);
+        for (const auto &field : record.getTagRange("LOK")) {
+            if (field.hasSubfieldWithValue('0', "935") and field.hasSubfieldWithValue('a', "iSWA")) {
+                found_iSWA = true;
+                break;
+            }
+        }
+        if (not found_iSWA)
+            continue;
+
+        const auto _773_field(record.findTag("773"));
+        if (_773_field == record.end())
+            continue;
+
+        const auto _773_subfields(_773_field->getSubfields());
+        std::string superior_ppn;
+        for (const auto &subfield : _773_subfields) {
+            if (subfield.code_ == 'w' and StringUtil::StartsWith(subfield.value_, "(DE-627)")) {
+                superior_ppn = subfield.value_.substr(__builtin_strlen("(DE-627)"));
+                break;
+            }
+        }
+        if (superior_ppn.empty())
+            continue;
+
+        const auto subfield_d(_773_subfields.getFirstSubfieldWithCode('d'));
+        static auto year_extractor(RegexMatcher::RegexMatcherFactoryOrDie(".*(\\d{4}).*"));
```

Comment:
.* at beginning and end of regex makes no sense & should be removed for readability+performance reasons

Reply:
maybe you could also remove the brackets () and use (*year_extractor)[0] on line 81 instead

Reply:
Addition: .*(\\4).* will also match 19202033 because .* also matches digits, so you might want to do something like [^0-9]*(\\d{4})[^0-9]*